Start containers in their own cgroup namespaces

rgulewich opened the pull request:
**- What I did**

This adds [cgroup namespace support](http://man7.org/linux/man-pages/man7/cgroup_namespaces.7.html) to docker. It's enabled for all containers that are not run with `--privileged`, if the kernel supports it.

Fixes #38332.

**- How I did it**

Now that [runc supports cgroup namespaces](https://github.com/opencontainers/runc/pull/1916), just add a namespace of type "cgroup" to the list.

At daemon startup, this looks for `/proc/self/ns/cgroup` to be present to determine whether or not to enable it.

**- How to verify it**

Run a new container and verify that its pid 1 is running in its own cgroup namespace.

**- Description for the changelog**

Add cgroup namespace support for Linux containers

**- A picture of a cute animal (not mandatory but encouraged)**

rgulewich committed:
Start containers in their own cgroup namespaces

This is enabled for all containers that are not run with --privileged,
if the kernel supports it.

Fixes #38332

Signed-off-by: Rob Gulewich <rgulewich@netflix.com>

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "38332-cgroup-ns" git@github.com:rgulewich/moby.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

thaJeztah commented on pkg/sysinfo/sysinfo_linux.go:
```diff
@@ -62,6 +62,11 @@ func New(quiet bool) *SysInfo {
 
+	// Check if cgroup namespaces are supported.
+	if _, err := os.Stat("/proc/self/ns/cgroup"); !os.IsNotExist(err) {
+		sysInfo.CgroupNamespaces = true
```
We may want to extract it to a function, e.g.

```suggestion
		sysInfo.CgroupNamespaces = cgroupNamespacesSupported()
```

Looking at that, perhaps this should be a utility in libcontainer; somewhere around here; https://github.com/opencontainers/runc/blob/f5b99917df9fbe1d9a4114966fb088dd6860e85a/libcontainer/cgroups/utils.go#L30-L45

and/or in the containerd/cgroups package https://github.com/containerd/cgroups/blob/master/utils.go

rgulewich replied:
Good idea - it makes sense to me for this to live in libcontainer.  What do you think about adding it in its own sub-package of libcontainer?  Everything in the `libcontainer/cgroups` package deals with cgroups themselves, whereas this is actually a namespace - I worry that putting them together in the cgroups package might make this more confusing.

So maybe something like `libcontainer/namespace/cgroups.go`?

thaJeztah commented on pkg/sysinfo/sysinfo_linux.go:
```diff
@@ -62,6 +62,11 @@ func New(quiet bool) *SysInfo {
 		}
 	}
 
+	// Check if cgroup namespaces are supported.
```
Would be good to move this a bit more up, below 

```go
 	_, ok := cgMounts["devices"]
 	sysInfo.CgroupDevicesEnabled = ok
```

thaJeztah commented on daemon/oci_linux.go:
```diff
@@ -309,6 +309,12 @@ func setNamespaces(daemon *Daemon, s *specs.Spec, c *container.Container) error
 	}
 
+	// cgroup
+	if daemon.cgroupNamespacesEnabled && !c.HostConfig.Privileged {
```
Wondering if this should be enabled by default if the host support it, or if it should be an opt-in/opt-out

I'm not too familiar with the feature itself, but my train of thought is; is this causing a change in behavior? Would there be situations where namespaced cgroups are not desirable?

@AkihiroSuda @crosbymichael ?

rgulewich replied:
I was on the fence about this myself - I listed a couple other options in [the GH issue](https://github.com/moby/moby/issues/38332).  @cpuguy83, any thoughts?

cyphar replied:
The main change in behaviour is that `/proc/self/cgroup` will now hide the Docker-related stuff. Personally I think it should be enabled by default and shouldn't be daemon-wide at all -- it should be treated like all other namespaces. My only worry is that sharing a cgroup with `--cgroup-parent` will now become more complicated because if you have two containers in the same cgroup but in different cgroup namespaces then things start to get a bit weird (especially with `nsdelegate` thrown in).

cpuguy83 replied:
Sorry, I missed this notification.

Agree we should enable by default, but make sure not to break users who expect things like cgroup-parent and privileged to work a certain way.

rgulewich replied:
The current behaviour is to enable it by default (just like any other namespace), except if the container is privileged.

The main side effect (which shows up in the integration tests), is that the cgroup hierarchy is no longer visible from inside the container.  I don't know if this qualifies as a breaking change or not.

@cyphar / @cpuguy83, what are your concerns about running with `--cgroup-parent`?  I can add some integration tests to confirm the behaviour.

cyphar replied:
> The main side effect (which shows up in the integration tests), is that the cgroup hierarchy is no longer visible from inside the container. I don't know if this qualifies as a breaking change or not.

This is the main purpose of cgroup namespaces, so it's good that it works. :wink: I don't think it'll count as a breaking change unless some folks have really tightly integrated into what `/proc/self/cgroups` contains which really isn't a good idea anyway.

> what are your concerns about running with `--cgroup-parent`?

Since writing the above comment I've remembered that cgroup sharing doesn't exist in Docker (getting two containers to share a cgroup) and so `--cgroup-parent` shouldn't be a problem after all. The concern would be that if you have two containers sharing a cgroup you would want them to share cgroup namespaces.

justincormack replied:
We do support sharing of most namespaces, so we could add support for cgroup namespace sharing.

thaJeztah commented on pkg/sysinfo/sysinfo.go:
```diff
@@ -16,6 +16,9 @@ type SysInfo struct {
 	cgroupPids
 
+	// Whether the kernel supports cgroup namespaces or not
+	CgroupNamespaces bool
```
Perhaps name this `...Enabled` (if it's `opt-in`) or `..Disabled` (if we'll make it `opt-out`) (see my other comment) to be consistent with other options

```suggestion
	CgroupNamespacesEnabled bool
```

Should this also be in a separate type, and embedded (as we did for the other features?)


edit: hm.. looks like we're not really consistent on the naming, so not sure

rgulewich replied:
I copied what was done for AppArmor and Seccomp - those were maybe poor examples to follow. :) I'm happy to break it out into an embedded type, if that's the preferred way of doing it.

thaJeztah reviewed (changes requested):
Thanks! Only gave this a cursory glance, but left some comments.

Looks like this information will be returned by the `/info` endpoint, so we'll need changes to the swagger file https://github.com/moby/moby/blob/master/api/swagger.yaml, and the API changelog https://github.com/moby/moby/blob/master/docs/api/version-history.md as well

thaJeztah commented:
CI failures look related

```
23:54:57 FAIL: docker_cli_build_test.go:3973: DockerSuite.TestBuildContainerWithCgroupParent
23:54:57 
23:54:57 docker_cli_build_test.go:3996:
23:54:57     c.Fatalf("There is no expected memory cgroup with parent /%s/: %s", cgroupParent, result.Combined())
23:54:57 ... Error: There is no expected memory cgroup with parent /test/: Sending build context to Docker daemon  2.048kB


23:54:57 Step 1/2 : FROM busybox
23:54:57  ---> 6ad733544a63
23:54:57 Step 2/2 : RUN cat /proc/self/cgroup
23:54:57  ---> Running in a799443f87c8
23:54:57 11:devices:/
23:54:57 10:pids:/
23:54:57 9:cpuset:/
23:54:57 8:cpu,cpuacct:/
23:54:57 7:freezer:/
23:54:57 6:net_cls,net_prio:/
23:54:57 5:hugetlb:/
23:54:57 4:blkio:/
23:54:57 3:perf_event:/
23:54:57 2:memory:/
23:54:57 1:name=systemd:/
23:54:57 Removing intermediate container a799443f87c8
23:54:57  ---> 48b98162374e
23:54:57 Successfully built 48b98162374e
23:54:57 Successfully tagged buildcgroupparent:latest
23:54:57 

00:15:14 FAIL: docker_cli_run_test.go:3237: DockerSuite.TestRunContainerWithCgroupParent
00:15:14 
00:15:14 docker_cli_run_test.go:3242:
00:15:14     // cgroup-parent relative path
00:15:14     testRunContainerWithCgroupParent(c, "test", "cgroup-test")
00:15:14 docker_cli_run_test.go:3267:
00:15:14     c.Fatalf("unexpected cgroup paths. Expected at least one cgroup path to have suffix %q. Cgroup Paths: %v", expectedCgroup, cgroupPaths)
00:15:14 ... Error: unexpected cgroup paths. Expected at least one cgroup path to have suffix "test/ee671df4af17e4f7594f6199410af9e09025b4cefb4f3fdef3391f35eaf976d5". Cgroup Paths: map[cpu,cpuacct:/ hugetlb:/ blkio:/ memory:/ name=systemd:/ perf_event:/ devices:/ pids:/ cpuset:/ freezer:/ net_cls,net_prio:/]
00:15:14 


00:16:09 FAIL: docker_cli_run_test.go:3272: DockerSuite.TestRunInvalidCgroupParent
00:16:09 
00:16:09 docker_cli_run_test.go:3276:
00:16:09     testRunInvalidCgroupParent(c, "../../../../../../../../SHOULD_NOT_EXIST", "SHOULD_NOT_EXIST", "cgroup-invalid-test")
00:16:09 docker_cli_run_test.go:3307:
00:16:09     c.Fatalf("unexpected cgroup paths. Expected at least one cgroup path to have suffix %q. Cgroup Paths: %v", expectedCgroup, cgroupPaths)
00:16:09 ... Error: unexpected cgroup paths. Expected at least one cgroup path to have suffix "SHOULD_NOT_EXIST/42c0cf9e9b4892b42a23b57f943d417a52a8373608c7d69b32e56bc528bc0761". Cgroup Paths: map[pids:/ net_cls,net_prio:/ blkio:/ perf_event:/ name=systemd:/ devices:/ cpuset:/ cpu,cpuacct:/ freezer:/ hugetlb:/ memory:/]
00:16:09 


00:24:18 FAIL: docker_cli_daemon_test.go:1857: DockerDaemonSuite.TestDaemonCgroupParent
00:24:18 
00:24:18 [d2f8a22850347] waiting for daemon to start
00:24:18 [d2f8a22850347] daemon started
00:24:18 
00:24:18 docker_cli_daemon_test.go:1881:
00:24:18     c.Assert(found, checker.True, check.Commentf("Cgroup path for container (%s) doesn't found in cgroups file: %s", expectedCgroup, cgroupPaths))
00:24:18 ... obtained bool = false
00:24:18 ... Cgroup path for container (test/0c465dd7a395dc6ce8c0d6a292233f3580b0192f20db92712fb892077c8c8268) doesn't found in cgroups file: map[devices:/ pids:/ cpuset:/ freezer:/ perf_event:/ memory:/ name=systemd:/ cpu,cpuacct:/ net_cls,net_prio:/ hugetlb:/ blkio:/]
00:24:18 
00:24:18 [d2f8a22850347] exiting daemon
00:24:18 [d2f8a22850347] waiting for daemon to start
00:24:18 [d2f8a22850347] daemon started
00:24:18 
00:24:18 [d2f8a22850347] exiting daemon
00:24:18 

```

rgulewich commented:
@thaJeztah This change doesn't change the `/info` endpoint on my machine:

```
$ curl -sS http://localhost:4243/v1.26/info | jq . | grep -i cgroup
  "CgroupDriver": "cgroupfs",
```

*Should* it be added to that endpoint?

olljanat commented:
@rgulewich plz check failing tests

olljanat commented:
Derek add label: rebuild/z

olljanat commented:
@rgulewich #38426 removed SameHostDaemon check so CI fails now to:
```
07:13:07 # github.com/docker/docker/integration-cli [github.com/docker/docker/integration-cli.test]
07:13:07 integration-cli/docker_cli_daemon_test.go:1858:18: undefined: SameHostDaemon
07:13:07 integration-cli/docker_cli_run_test.go:3239:18: undefined: SameHostDaemon
```
so can your rebase and update tests?

rgulewich commented:
@olljanat Pushed changes.  The experimental failures don't look to be related, unless I'm mistaken.

rgulewich commented:
Would anyone be able to give a more thorough review of this?  I agree with @cyphar's comment in https://github.com/opencontainers/runc/pull/1976 that moving the detection code into libcontainer isn't really necessary (though that PR is there if that's preferable).  @cpuguy83, would you mind checking that any concerns you have are covered?

Thanks!

cpuguy83 commented:
I'm a bit worried that we are having to change tests and as such making a breaking change... but maybe the tests are just not good tests.
I feel like the cgroup API is pretty well defined here and depending on a particular heirarchy for a cgroup seems wrong.... but still worried that people could be doing this, particularly because there's no way to turn it off.

rgulewich commented:
@cpuguy83 - Fair enough.  I wouldn't want to break someone who was depending on this behaviour.  Would a run-time flag to enable / disable this for the entire daemon be sufficient, do you think?

thaJeztah commented:
Could be a separate flag, or perhaps use the new `features` option in `daemon.json` (we were discussing to add this as a flag as well); see https://github.com/moby/moby/pull/37593 and https://github.com/moby/moby/pull/37692)

> It's enabled for all containers that are not run with `--privileged`, if the kernel supports it

do we want to have an option to enable/disable it separately (per container)? (e.g. `--security-opt <something>`). Basically, `--privileged` is already too coarse-grained, so I'd like to prevent people from "just set `--privileged` and it works", because that's throwing away _all_ security that containers provides.

thaJeztah commented:
ping @justincormack @crosbymichael @AkihiroSuda @kolyshkin would love to have your input on this one

justincormack commented:
Hmm, need to think about the compatibility change. Another option would be to enable by default but to have a daemon wide option to disable. I think for almost all users the change won't matter but there seem to be people doing all sorts of weird things.

It would be most consistent to allow `--cgroupns=host` (and maybe other containers) as a way to remove for a single container.

kolyshkin commented:
:+1:  for `--cgroupns=host`, that would be very in line with others (i.e. `--network`, `--ipc`, `--pid`, `--userns`, `--uts`...)

rgulewich committed:
Make cgroup namespaces configurable

This adds both a daemon-wide flag and a container creation property:
- Set the `CgroupnsMode: "host|private"` HostConfig property at
  container creation time to control what cgroup namespace the container
  is created in
- Set the `--default-cgroupns-mode=host|private` daemon flag to control
  what cgroup namespace containers are created in by default
- Set the default if the daemon flag is unset to "host", for backward
  compatibility
- Default to CgroupnsMode: "host" for client versions < 1.40

Signed-off-by: Rob Gulewich <rgulewich@netflix.com>

rgulewich commented:
@kolyshkin / @justincormack / @cpuguy83 / @crosbymichael / @AkihiroSuda  - Mind taking a look?  I made the change to make cgroup namespaces configurable via both a daemon-wide flag and a container creation property (similar to how `IpcMode` works):

- Added a `CgroupnsMode: "host|private"` HostConfig property at container creation time to control what cgroup namespace the container is created in
- Added a `--default-cgroupns-mode=host|private` daemon flag to control what cgroup namespace containers are created in by default
- The daemon default (if `--default-cgroupns-mode` is unset) is "host", for backward compatibility

(The `make validate` errors appear to be unrelated to my change, but I'll take a look)

thaJeztah commented:
Saw your comment fly by in my notifications; Just a very quick comment before I forget; make sure that the behaviour for containers created with older API versions does not change; i.e. they should _not_ use `private` (even if the daemon has it configured as default), and the `CgroupnsMode` options should be ignored; see https://github.com/moby/moby/pull/38700 / https://github.com/moby/moby/pull/35621

rgulewich commented:
> Saw your comment fly by in my notifications; Just a very quick comment before I forget; make sure that the behaviour for containers created with older API versions does not change; i.e. they should _not_ use `private` (even if the daemon has it configured as default), and the `CgroupnsMode` options should be ignored; see #38700 / #35621

@thaJeztah - Good catch, and thanks for the pointer to the other issues.  Updated so that versions < 1.40 default to `private`, regardless of the daemon default.

rgulewich commented:
@kolyshkin / @justincormack / @cpuguy83 / @crosbymichael / @AkihiroSuda - Giving this a bump.  Would any of you mind reviewing?

justincormack commented:
Needs a rebase.

rgulewich commented:
> Needs a rebase.

@justincormack - Thanks; done.

rgulewich commented:
@kolyshkin / @justincormack / @cpuguy83 / @crosbymichael / @AkihiroSuda - any of you folks mind taking a look at this?  Thanks!

thaJeztah commented:
ping @justincormack PTAL

kolyshkin commented:
SGTM

AkihiroSuda commented:
Can we keep the host mode as the default regardless to API version for minimizing the risk of causing incompatibility?

rgulewich commented:
> Can we keep the host mode as the default regardless to API version for minimizing the risk of causing incompatibility?

@AkihiroSuda - Yes, the default (unless overridden with the `--default-cgroupns-mode` command-line option) is "host" - see `DefaultCgroupNamespaceMode ` in `daemon/config/config_unix.go`.

AkihiroSuda commented:
LGTM except CI failure

rgulewich commented:
@AkihiroSuda - Fixed the CI issue.  The failing experimental tests are `DockerSwarmSuite.TestAPISwarmLeaderElection` and `DockerSwarmSuite.TestSwarmClusterRotateUnlockKey`, which don't seem to be related to my changes.

@justincormack, would you mind giving this a review as well?  Thanks!

cpuguy83 reviewed (approved):
LGTM

rgulewich commented:
@thaJeztah - Is there anything else I can do on my end to nudge this toward being merged?  Thanks!

yongtang commented:
All tests passed, merging.

AkihiroSuda commented:
Is there docker CLI pr for overriding `default-cgroupns-mode` ?

thaJeztah commented:
@akihirosuda I don't think there is one yet (at least; I don't see one linked)

tonistiigi commented:
@rgulewich `TestCgroupNamespacesRunPrivileged` added in this test seems to be failing consistently with 

`assertion failed: error is not nil: Error response from daemon: privileged mode is incompatible with private cgroup namespaces.  You must run the container in the host cgroup namespace when running privileged mode`

eg.  https://jenkins.dockerproject.org/job/Docker%20Master/label=centos-7-noselinux-devicemapper/10631/consoleFull

It doesn't fail in PRs and in master just because it is skipped because of the system condition checks. Can you follow-up please?

rgulewich commented:
> @rgulewich `TestCgroupNamespacesRunPrivileged` added in this test seems to be failing consistently with
> 
> `assertion failed: error is not nil: Error response from daemon: privileged mode is incompatible with private cgroup namespaces. You must run the container in the host cgroup namespace when running privileged mode`
> 
> eg. https://jenkins.dockerproject.org/job/Docker%20Master/label=centos-7-noselinux-devicemapper/10631/consoleFull
> 
> It doesn't fail in PRs and in master just because it is skipped because of the system condition checks. Can you follow-up please?

Good catch, @tonistiigi - please see my PR here: https://github.com/moby/moby/pull/39578
